[회원 가입] 회원가입 연관 비즈니스 로직 & API 구현(#8)

Minjae-An opened the pull request:
## 작업 대상

<!-- 작업한 대상을 자세히 설명해주세요. -->

- 회원 가입 연관 비즈니스 로직 및 API

## 📄 작업 내용

<!-- 작업한 내용을 간단히 요약해주세요. -->

- access token, refresh token 생성/검증 로직 구현
- 비밀번호 암호화시 사용되는 salt 생성 로직 구현
- 회원가입 API 구현
- 이외 관련 엔티티(지역, 동물상 등) 조회 로직 구현

## 🙋🏻 주의 사항

<!-- 리뷰어를 위해 복잡하거나 중요한 코드를 명시해주세요. -->

- 32바이트의 랜덤 바이너리를 base64 인코딩하여  `salt` 생성
- 회원 엔티티 생성, 프로필 사진 엔티티 생성, 회원 선호조건 엔티티 생성은 하나의 트랜잭션으로 다뤄져야 한다고 생각
- 이에 따라 `UserService` 에 엔티티를 생성하는 로직을 구현하고 `JoinService` 에서 해당 로직들을 하나의 트랜잭션으로 묶어 처리하도록 구성

## 📎 관련 이슈

- 체형, 동물상, 음역대도 enum으로 다룰 필요가 있을지?

## 레퍼런스

<!-- 작업 시 참고했던 문건의 URL을 남겨주세요 -->

- https://stackoverflow.com/questions/184112/what-is-the-optimal-length-for-user-password-salt

Minjae-An committed:
feat: 화살 보내기 요청 DTO 추가(#24)

보낼 화살 수를 담은 요청 DTO 정의

Minjae-An committed:
feat: 확살 송수신 내역 repository, 내역 확인 로직 추가(#24)

- 화살 송수신 내역 repository 정의
- 보낸 사용자, 받은 사용자를 통해 내역 존재 여부를 확인하는 로직 구현

Minjae-An committed:
feat: 화살 송수신 내역 존재 여부 확인 로직 추가(#24)

보낸 사용자, 받은 사용자를 통해 송수신 내역 존재 여부를 확인하는
로직 구현

Minjae-An committed:
feat: 화살 송수신 내역 저장 기능 구현(#24)

Minjae-An committed:
feat: 화살 관련 예외 enum 정의(#24)

다음 상황들에 대한 예외 enum을 정의하였다.
- 자기 자신에게 화살을 보내는 경우
- 이미 화살을 보낸 사용자인 경우
- 가진 화살 수가 부족해 화살을 보낼 수 없는 경우

Minjae-An committed:
feat: 사용자 화살 감소 로직 추가(#24)

감소하려는 화살 수가 사용자가 보유한 화살 수보다 클 경우 예외 발생

Minjae-An committed:
refactor: 생성 시간 필드, 불필요한 업데이트 불가 DDL 속성 삭제(#24)

Minjae-An committed:
feat: 화살 보내기 비즈니스 로직 추가(#24)

화살 보내기 비즈니스 로직은 다음 과정을 거친다.
1. 자기 자신에게 화살을 보내는 상황 검증
2. 화살을 보낸 사용자에게 또 보내는 상황 검증
3. 화살 내역 저장
4. 보내는 사용자 화살 감소, 화살이 부족할 경우 예외 발생(과정 3 롤백)
5. 받는 사용자 화살 증가

Minjae-An committed:
feat: 화살 보내기 API 구현(#24)

Minjae-An committed:
refactor: 화살 command repository 삭제(#24)

- query, command를 service, repository를 중개하는 별도의 계층을 통해 구분
- 이에 따라 기존 command repository 삭제
- 기존 command repository 사용 위치, 화살 command로 대체

Minjae-An committed:
refactor: 코드 정렬 수정(#24)

Minjae-An committed:
Merge branch 'dev' into feat/#24-send-arrow

# Conflicts:
#	be/src/main/java/yeonba/be/arrow/repository/ArrowCommand.java
#	be/src/main/java/yeonba/be/arrow/service/ArrowService.java
#	be/src/main/java/yeonba/be/exception/CommonException.java
#	be/src/main/java/yeonba/be/user/entity/User.java

Minjae-An committed:
refactor: 사용자 화살 증감 메서드 이름 수정(#24)

서로 대치되는 작업을 수행하는 상황을 잘 표현할 수 있도록 메서드 이름 수정

Minjae-An committed:
feat: 화살 관련 예외 enum 추가(#24)

- 이미 화살을 보낸 사용자 예외
- 화살이 부족하여 화살을 보낼 수 없는 예외

Minjae-An committed:
refactor: 화살 내역 존재 여부 확인 메서드 이름 수정(#24)

가독성 향상을 위해 간단한 이름을 수정

Minjae-An committed:
refactor: 코드 배치 수정(#24)

dev 브랜치 작업 내역 병합에 따른 코드 배치 수정

Minjae-An committed:
refactor: 화살 보내기 API 파라미터 이름 수정(#24)

좀 더 명확하게 대치되는 의미를 나타낼 수 있도록 화살을 받는
사용자 ID 파라미터 이름을 receiverId로 수정

Minjae-An committed:
refactor: 컨벤션에 맞게 코드 배치 수정(#24)

Minjae-An committed:
feat: 이름 칼럼, 생성자 추가(#27)

- 이름 칼럼 추가
- 테스트시 활용 가능한 생성자 추가

Minjae-An committed:
feat: 상대 프로필 응답 DTO 수정(#27)

- 래퍼 타입 필드들 기본 타입으로 변경
- 프로필 사진 URL 리스트 필드 추가
- 사용자가 가진 총 화살 수 필드 추가
- 음주 성향, 흡연 성향 필드 추가
- 코드 정렬

Minjae-An committed:
feat: 사용자 엔티티 수정(#27)

- 레퍼런스 타입 not null 필드들, Column 어노테이션 사용 명시
- salt 필드 추가
- 생성 일시, 최종 수정 일시 필드 추가
- JpaAuditing 활용, 생성/최종 수정 일시 필드들 엔티티 저장시 자동 설정
- 새로운 생성자 구성

Minjae-An committed:
feat: 상대 프로필 조회 비즈니스 로직 추가(#27)

프로필을 조회하는 사용자 ID의 경우 상대방이 화살을 보넀던 사용자인지
확인하기 위해 파라미터로 받는다.

Minjae-An committed:
feat: 상대 프로필 API 구현 및 코드 정렬(#27)

Minjae-An committed:
feat: 사용자 엔티티 수정(#27)

- 요구사항 변경에 따라 음주 성향, 흡연 성향 필드 삭제
- 성별 정보 제공 메서드 추가

Minjae-An committed:
feat: 사용자 프로필 응답 DTO 수정(#27)

- 음주 성향, 흡연 성향 필드 삭제
- 성별 필드 추가

Minjae-An committed:
feat: 사용자 프로필 상세 응답 DTO 수정(#27)

- 음주 성향, 흡연 성향 필드 삭제
- 생성자 성별 필드 설정 로직 수정
- 코드 정렬

Minjae-An committed:
feat: 사용자 프로필 조회 비즈니스 로직 수정(#27)

- 음주 성향, 흡연 성향 삭제
- 성별 정보 응답에 포함토록 수정
- 코드 정렬

Minjae-An committed:
chore: spring mail 종속성 추가(#35)

Minjae-An committed:
chore: Gmail SMTP 사용을 위한 설정 추가(#35)

Minjae-An committed:
feat: 이메일 기반 사용자 조회 로직 추가(#35)

Minjae-An committed:
feat: 이메일 기반 사용자 조회 로직 추가(#35)

Minjae-An committed:
feat: 서비스 사용 정규식 enum 정의(#35)

- 서비스에서 사용하는 정규식들을 모아놓는 enum 클래스 정의
- 이메일, 비밀번호 정규식 정의

Minjae-An committed:
feat: 임시 비밀번호 발급 기능 구현(#35)

- 비밀번호 정규식을 충족하는 임의의 임시 비밀번호를 발급
- Random보다 생성하는 난수를 예측하기 어려워 보안적으로 뛰어난 SecureRandom 사용

Minjae-An committed:
feat: 메일 관련 설정 정의(#35)

- spring mail에서 이메일 전송을 위해 제공하는 JavaMailSender 빈 등록
- 필요한 설정 값들은 application.yml에 정의된 값들을 끌어와 사용
- 빈 등록 로직에서 구글 SMTP 서버를 통해 메일을 전송하기 위한 설정 코드 추가

Minjae-An committed:
feat: 이메일 전송 로직 추가(#35)

- 단순 텍스트 이메일 전송 기능 구현
- 발신자는 application.yml에 정의해논 구글 SMTP 사용자 이름 값으로 설정

Minjae-An committed:
feat: 임시 비밀번호 발급 비즈니스 로직 추가(#35)

임시 비밀번호는 다음 과정을 거친다.
1. 요청 이메일 기반 사용자 조회
2. 임시 비밀번호 생성
3. 사용자 비밀번호, 임시 비밀번호로 변경
4. 임시 비밀번호 발급 메일 전송
임시 비밀번호를 암호화하는 로직은 추후 구현 예정

Minjae-An committed:
feat: 임시 비밃번호 발급 비즈니스 로직 트랜잭션 정의 추가(#35)

Minjae-An committed:
feat: 임시 비밀번호 발급 요청 DTO 수정(#35)

- 불필요한 기존 생성자 삭제, 기본 생성자 추가
- 비밀번호 형식 제약 조건 추가
- 코드 정렬

Minjae-An committed:
feat: 임시 비밀번호 발급 API 구현(#35)

- 임시 비밀번호 발급 API 구현
- 명세 속성 수정
- 코드 정렬

Minjae-An committed:
Merge branch 'dev' into feat/#35-send-temporary-password

# Conflicts:
#	be/src/main/java/yeonba/be/arrow/service/ArrowService.java
#	be/src/main/java/yeonba/be/user/controller/UserController.java
#	be/src/main/java/yeonba/be/user/entity/User.java

Minjae-An committed:
Merge branch 'dev' into feat/#35-send-temporary-password

# Conflicts:
#	be/src/main/java/yeonba/be/user/entity/User.java

Minjae-An committed:
Merge branch 'dev' into feat/#35-send-temporary-password

Minjae-An committed:
refactor: 병합에 따른 코드 재배치(#35)

dev 브랜치 작업 내역 병합에 따른 코드 재배치

Minjae-An committed:
refactor: 병합에 따른 코드 재배치(#35)

dev 브랜치 작업 내역 병합에 따른 코드 재배치

Minjae-An committed:
chore: sms 서비스 sdk 의존성 추가(#28)

coolsms 서비스에서 제공하는 java sdk 의존성 추가

Minjae-An committed:
chore: sms 서비스 관련 설정 추가(#28)

sms 서비스를 이용하기 위해 필요한 설정들 추가

Minjae-An committed:
feat: sms 서비스 이용을 위한 설정 정의(#28)

- coolsms에서는 메시지 전송을 수행할 수 있는 DefaultMessageService 제공
- config에서 application.yml을 통해 값을 주입 받아 DefaultMessageService 빈 등록

Minjae-An committed:
feat: 인증 코드 생성 로직 추가(#28)

- apache.common.lang3 라이브러리의 RandomStringUtils 이용
- 영어 대소문자, 숫자로 구성된 6자리 랜덤 인증 코드를 생성
- 영어 대소문자 52개(26+26), 숫자 10개로 총 62개의 선택 가능 문자 존재
- 생성 가능한 인증 코드 수, 약 56억 8천만 개(62^6)로 보안성을 확보하였음

Minjae-An committed:
feat: 문자 전송 로직 추가(#28)

Minjae-An committed:
feat: 전화번호 기반 사용자 존재 여부 확인 로직 정의(#28)

Minjae-An committed:
feat: 전화번호 기반 사용자 존재 여부 확인 로직 추가(#28)

Minjae-An committed:
feat: 전화번호 형식 제약 조건 추가(#28)

010으로 시작하며 0~9까지의 숫자로 이뤄진 11자리 문자열만 허용토록 제약 조건 추가

Minjae-An committed:
chore: application.yml 수정(#28)

sms.provider 속성도 application.properties 통해 관리하도록 수정

Minjae-An committed:
feat: 인증 코드 sms 전송 비즈니스 로직 추가(#28)

- 요청된 전화번호를 통해 사용자 존재 확인
- 사용자가 존재할 경우만 sms로 인증 코드 전송

Minjae-An committed:
feat: 전화번호 인증 코드 전송 API 구현(#28)

Minjae-An committed:
chore: spring data redis 의존성 추가(#28)

Minjae-An committed:
chore: redis 관련 설정 추가(#28)

Minjae-An committed:
feat: redis 관련 설정 정의(#28)

- redis 커넥션 팩토리 빈 등록
- CRUD 연산 지원을 위한 RedisTemplate 빈 등록

Minjae-An committed:
feat: redis 키-값 저장,조회,삭제 로직 추가(#28)

- redis를 이용한 저장, 삭제 기능 지원 유틸 클래스 정의
- 만료 시간을 분 단위로 설정하는 저장 로직 구현
- 기본 조회 로직 구현, Optional 리턴 타입을 통해 null 처리 지원
- 기본 삭제 로직 구현

Minjae-An committed:
feat: 인증 코드 sms 전송 비즈니스 로직 수정(#28)

- 인증 코드 sms 전송 이전 redis에 번호를 키로 코드를 저장하는 로직 추가
- 새 인증 코드 발급 이전 기존 발급 내역 삭제

Minjae-An committed:
feat: 인증 코드 정규식 정의(#28)

영어 대소문자, 숫자로 이뤄진 6자리 문자열만 허용하는 정규식 enum 정의

Minjae-An committed:
feat: 아이디 찾기 예외 enum 정의(#28)

- 인증 코드 내역이 존재하지 않는 경우(인증 코드 만료도 포함)
- 인증 코드가 일치하지 않는 경우

Minjae-An committed:
feat: 전화번호 기반 사용자 조회 기능 정의(#28)

Minjae-An committed:
feat: 전화번호 기반 사용자 조회 로직 추가(#28)

Minjae-An committed:
refactor: 코드 정렬(#28)

Minjae-An committed:
feat: 아이디 찾기 요청 DTO 수정(#28)

- 인증 번호를 발급한 전화번호 필드 추가
- 인증 코드 필드 명세 속성 수정
- 인증 코드 정규식 제약 조건 추가
- 불필요한 기존 생성자, 기본 생성자로 대체

Minjae-An committed:
feat: 아이디 찾기 비즈니스 로직 추가(#28)

아이디 찾기 요청은 인증 코드 발급 전화번호, 인증 코드를 포함한다.
아이디 찾기는 다음 과정을 거쳐 이뤄진다.
1. 발급 전화번호를 기반으로 인증 코드 redis에서 조회
2. 인증 코드 일치 확인
3. 전화번호 기반 사용자 조회
4. 확인한 인증 코드 삭제
5. 찾은 이메일을 응답으로 반환

Minjae-An committed:
feat: 아이디 찾기 API 구현(#28)

Minjae-An committed:
feat: 전화번호 정규식 정의(#28)

Minjae-An committed:
feat: 회원가입 요청 DTO 수정(#8)

- 요구사항 변경에 따라 필드들 삭제 및 추가
- 필드들 API 명세
- 문자열 필드들 정규식 제약 조건 추가
- 불필요한 생성자 삭제, 기본 생성자로 대체

Minjae-An committed:
Merge branch 'feat/#28-verification-code-to-phone' into feat/#8-join-api

# Conflicts:
#	be/src/main/java/yeonba/be/mypage/dto/response/UserProfileDetailResponse.java
#	be/src/main/java/yeonba/be/user/dto/response/UserProfileResponse.java
#	be/src/main/java/yeonba/be/user/entity/User.java

Minjae-An committed:
refactor: PasswordEncryptor 최상위 util 패키지로 이동(#8)

- 비밀번호 암호화 기능을 전역적으로 사용하기에 패키지 경로 변경
- 사용 위치 수정

Minjae-An committed:
refactor: 잘못된 들여쓰기 수정을 위해 코드 정렬(#8)

Minjae-An committed:
feat: 이름 기반 동물상 조회 로직 정의(#8)

Minjae-An committed:
feat: 이름 기반 닮은 동물상 조회 로직 구현(#8)

Minjae-An committed:
feat: 이름 기반 지역 조회 로직 정의(#8)

Minjae-An committed:
feat: 이름 기반 지역 조회 로직 구현(#8)

Minjae-An committed:
feat: 분류 기반 음역대 조회 로직 정의(#8)

Minjae-An committed:
feat: 분류 기반 음역대 조회 로직 구현(#8)

Minjae-An committed:
feat: 사용자 선호 조건 repository 정의(#8)

Minjae-An committed:
feat: 사용자 선호 조건 엔티티 저장 로직 구현(#8)

Minjae-An committed:
feat: 사용자 선호조건 엔티티 구성(#8)

이성 추천, 다른 사용자 검색시 사용되는 사용자 선호조건 엔티티 정의
다음 필드들을 포함한다.
- 선호하는 나이 하한, 상한
- 선호하는 키 하한, 상한
- 선호하는 MBTI
- 선호하는 체형
- 선호하는 음역대
- 선호하는 지역
- 선호하는 동물상
- 선호 조건 소유 사용자

Minjae-An committed:
feat: 사용자 엔티티 저장 로직 구성(#8)

Minjae-An committed:
feat: 프로필 사진 repository 정의(#8)

Minjae-An committed:
feat: 프로필 사진 엔티티 저장 로직 구성(#8)

Minjae-An committed:
feat: 회원가입 관련 예외 enum 정의(#8)

다음 예외 enum 들을 정의
- 비밀번호와 비밀번호 확인 값 불일치 상황
- 존재하지 않는 음역대
- 존재하지 않는 동물상
- 존재하지 않는 지역

Minjae-An committed:
refactor: 분류 기반 음역대 조회 로직 수정(#8)

- 잘못된 인덴트 수정, 코드 정렬
- 회원가입 관련 예외 enum을 사용토록 로직 수정

Minjae-An committed:
refactor: 이름 기반 닮은 동물상 조회 기능 수정(#8)

- 잘못된 들여쓰기 수정, 코드 정렬
- 회원가입 관련 예외 enum을 이용하도록 수정

Minjae-An committed:
refactor: 이름 기반 지역 조회 로직 수정(#8)

- 잘못된 들여쓰기, 코드 정렬 수정
- 회원가입 관련 예외 enum 사용토록 수정

Minjae-An committed:
refactor: 잘못된 들여쓰기, 코드 정렬로 수정(#8)

Minjae-An committed:
refactor: 회원가입 요청 DTO 수정(#8)

- 기존 생성자 삭제, @ModelAttribute 사용을 위한 생성자 추가
- 잘못된 들여쓰기 수정을 위한 코드 정렬 수정

Minjae-An committed:
feat: 회원가입 응답 DTO 구성(#8)

회원가입시 바로 서비스를 이용할 수 있도록 access token과
refresh token을 바로 지급하는 형태로 구성

Minjae-An committed:
feat: 성별 enum 구성(#8)

- 성별 enum 정의
- 문자열 성별 값을 enum으로 치환할 수 있는 편의 메서드 구현

Minjae-An committed:
feat: 비밀번호 암호화시 사용되는 salt 생성 로직 구현(#8)

Minjae-An committed:
refactor: 잘못된 들여쓰기 수정, 코드 정렬(#8)

Minjae-An committed:
feat: 회원가입 관련 로직 추가(#8)

- 사용자 엔티티 생성 및 저장 로직
- 프로필 사진 엔티티 생성 및 저장 로직
- 사용자 선호조건 엔티티 생성 및 저장 로직

Minjae-An committed:
feat: 회원가입 비즈니스 로직 초안 구성(#8)

회원가입 비즈니스 로직은 다음 과정을 거쳐 이뤄진다.
1. 사용자 엔티티 생성 및 저장
2. 프로필 사진 엔티티 생성 및 저장
3. 사용자 선호조건 엔티티 생성 및 저장
4. access token, refresh token 발급

Minjae-An committed:
feat: 회원가입 API 초안 구현 및 코드 정렬(#8)

- 회원가입 API 구현
- 잘못된 들여쓰기 수정을 위한 코드 정렬

Minjae-An committed:
refactor: 분류 기반 음역대 조회 로직 이름 수정, 코드 정렬(#8)

Minjae-An committed:
feat: 프로필 사진 엔티티 저장 로직 수정(#8)

무조건 일괄적 List 형태로 저장되는 특성을 고려해 List 파라미터로 받아
일괄 저장하도록 로직 수정

Minjae-An committed:
feat: S3 이용 프로필 사진 업로드 로직 추가(#8)

프로필 사진 업로드는 다음 과정을 거쳐 이뤄진다.
1. 사진 파일들 확장자 검증
2. 파일 식별을 위한 키 생성, profilephoto/{userId}-{photo idx} 형식
3. 사진 파일 업로드 요청 생성 및 업로드 수행

Minjae-An committed:
feat: 프로필 사진 엔티티 저장 로직에 S3 업로드 로직 추가(#8)

Minjae-An committed:
feat: 프로필 사진 컬렉션 필드 초기화 & 연관관계 설정 로직 추가(#8)

- NPE 방지를 위한 컬렉션 필드 초기화 설정
- 순수한 객체 관계를 고려한 프로필 사진 엔티티 연관관계 설정 로직 구성

Minjae-An committed:
feat: 이미 사용 중인 닉네임, 이메일 확인 로직 정의(#8)

- 이미 사용중인 닉네임 확인 로직 정의
- 이미 사용중인 이메일 확인 로직 정의

Minjae-An committed:
feat: 이미 사용 중인 닉네임, 이메일 확인 로직 추가(#8)

Minjae-An committed:
feat: 랜덤 salt 생성 로직 수정(#8)

- salt의 길이 고정 32비트(4바이트)로 변경
- 32비트 길이 salt 만으로 대부분의 보안적 위협 충분히 커버 가능
- base64 인코딩에 따라 로직에서 생성되는 salt 문자열의 길이는 8

Minjae-An committed:
feat: 이미 사용 중인 이메일/닉네임 예외 추가(#8)

- 이미 사용 중인 이메일 예외 enum 정의
- 이미 사용 중인 닉네임 예외 enum 정의

Minjae-An committed:
feat: s3 업로드된 프로필 사진 삭제 로직 추가(#8)

Minjae-An committed:
feat: 회원가입 관련 로직 수정(#8)

- 사용자 엔티티 저장 로직에 이미 사용 중인 이메일/닉네임 검증 로직 추가
- 프로필 사진 저장 로직에 사용자 엔티티와의 연관관계 설정 로직 추가

Minjae-An committed:
feat: 프로필 사진 삭제 이벤트 정의(#8)

- 회원가입 과정중 트랜잭션 롤백 발생시 업로드된 프로필 사진 삭제를 위해 정의
- 업로드된 프로필 사진 엔티티를 필드를 통해 전달

Minjae-An committed:
feat: 프로필 사진 삭제 이벤트 리스너 정의(#8)

- 회원가입 과정중 트랜잭션 롤백 발생시 프로필 사진 삭제 이벤트 발생
- 이벤트를 통해 삭제 대상인 프로필 사진 엔티티들을 전달 받음
- S3 서비스를 통해 해당 엔티티들의 삭제 작업 진행

Minjae-An committed:
feat: 회원가입 로직 롤백시 업로드된 프로필 사진 삭제 이벤트 발행 로직 추가(#8)

회원가입 도중 예외 상황 등으로 인해 트랜잭션 롤백시 이미 s3에 업로드된 프로필
사진들을 삭제하기 위한 이벤트 발행 로직 추가

Minjae-An committed:
chore: jjwt 의존성 추가(#8)

jwt 활용을 위해 사용되는 jjwt 라이브러리 의존성 추가

Minjae-An committed:
feat: access/refresh token 발급 및 검증 로직 추가(#8)

- jwt 시크릿 키 값의 경우 무작위 64바이트 바이너리를 16진수로 인코딩한 문자열 사용
- jwt 포맷의 access/refresh token 발급 로직 추가
- 만료 시간을 계산하는 공통 로직, 별도의 메서드로 추출
- 사용자 엔티티를 기반으로 jwt를 생성하는 공통 로직, 별도의 메서드로 추출
- access token의 보편적인 존속기간은 수 시간가량 따라서 8시간으로 설정
- refresh token의 보편적인 존속기간은 수 일에서 수 개월 내외 따라서 10일로 설정
- jwt의 일반적인 관행에 따라 사용자 이메일을 subject로 설정
- 사용자 엔티티의 PK를 클레임으로 설정하여 인가시 사용할 수 있도록 구성
- 검증 로직에선 시그니처의 유효성과 JWT 만료 여부를 검증

Minjae-An committed:
feat: refresh token 필드 및 업데이트 로직 추가(#8)

- 사용자의 refresh token을 저장할 필드 추가
- 사용자 refresh token 업데이트 로직 추가

Minjae-An committed:
feat: 회원가입 로직, access/refresh token 발급 로직 추가(#8)

- 회원가입 정상 완료의 결과로 access token과 refresh token 제공
- 회원가입 비즈니스 로직에 access token, refresh token 발급 로직 추가
- 새 refresh token 발급시 사용자 refresh token 업데이트 수행

Minjae-An committed:
refactor: 잘못된 주석 들여쓰기 수정(#8)

Minjae-An committed:
refactor: 문자열 값 Geneder enum 변환 메서드 이름 수정(#8)

Minjae-An committed:
refactor: 회원가입 비즈니스 로직 코드 정렬(#8)

Minjae-An committed:
refactor: Gender 문자열 값 변환 메서드 사용 위치 수정 및 코드 정렬(#8)

Minjae-An committed:
feat: salt 길이 32바이트로 변경(#8)

보안적으로 안전한 32자로 salt 길이 변경

Minjae-An committed:
Merge branch 'dev' into feat/#8-join-api

# Conflicts:
#	be/build.gradle
#	be/src/main/java/yeonba/be/config/MailConfig.java
#	be/src/main/java/yeonba/be/login/controller/LoginController.java
#	be/src/main/java/yeonba/be/login/dto/request/UserPasswordInquiryRequest.java
#	be/src/main/java/yeonba/be/login/service/LoginService.java
#	be/src/main/java/yeonba/be/user/entity/User.java
#	be/src/main/java/yeonba/be/user/repository/UserQuery.java
#	be/src/main/java/yeonba/be/user/repository/UserRepository.java
#	be/src/main/java/yeonba/be/util/EmailService.java
#	be/src/main/java/yeonba/be/util/TemporaryPasswordGenerator.java
#	be/src/main/resources/application.yml

Minjae-An committed:
refactor: 병합으로 코드 수정(#8)

Minjae-An commented:
불필요한 redis 관련 설정, 앞선 PR 내역 통합하며 삭제 예정

Minjae-An committed:
feat: 임시 비밀번호 발급 로직에 비밀번호 암호화 로직 추가(#8)

Minjae-An committed:
refactor: 잘못된 들여쓰기 간격 수정(#8)

acceptor-gyu reviewed (commented):
간격 확인해서 다시 올려주시기 바랍니다. 이 PR은 close 하겠습니다.

acceptor-gyu commented:
간격 확인해서 다시 올려주시기 바랍니다. 이 PR은 close 하겠습니다.

Minjae-An committed:
Merge branch 'dev' into feat/#8-join-api

# Conflicts:
#	be/build.gradle
#	be/src/main/java/yeonba/be/config/SmsConfig.java
#	be/src/main/java/yeonba/be/exception/LoginException.java
#	be/src/main/java/yeonba/be/login/controller/LoginController.java
#	be/src/main/java/yeonba/be/login/dto/request/UserIdInquiryRequest.java
#	be/src/main/java/yeonba/be/login/dto/request/UserPhoneNumberVerifyRequest.java
#	be/src/main/java/yeonba/be/login/dto/response/UserEmailInquiryResponse.java
#	be/src/main/java/yeonba/be/login/service/LoginService.java
#	be/src/main/java/yeonba/be/mypage/service/MyPageService.java
#	be/src/main/java/yeonba/be/user/entity/User.java
#	be/src/main/java/yeonba/be/user/repository/UserQuery.java
#	be/src/main/java/yeonba/be/user/repository/UserRepository.java
#	be/src/main/java/yeonba/be/util/ServiceRegex.java
#	be/src/main/java/yeonba/be/util/SmsService.java
#	be/src/main/java/yeonba/be/util/VerificationCodeGenerator.java

Minjae-An committed:
refactor: 병합으로 인한 코드 수정(#8)

dev 브랜치 내역 병합으로 인한 코드 수정

acceptor-gyu commented on be/src/main/java/yeonba/be/mypage/dto/response/UserProfileDetailResponse.java:
```diff
@@ -64,12 +64,12 @@ public class UserProfileDetailResponse {
         type = "string",
         description = "체형",
-        example = "마른 체형")
+        example = "메른 체형")
```
?

Minjae-An replied:
죄송합니다. 수정하겠습니다!

acceptor-gyu commented on be/src/main/java/yeonba/be/mypage/dto/response/UserProfileDetailResponse.java:
```diff
@@ -64,12 +64,12 @@ public class UserProfileDetailResponse {
         description = "사진 싱크로율",
         example = "80")
-    private int photoSyncRate;
+    private double photoSyncRate;
```
싱크로율 데이터는 integer 타입으로 응답, 저장 등 관리되는데 double로 바꾸신 이유가 무엇인가요?

Minjae-An replied:
업데이트하며 미처 수정되지 못한 부분 같습니다. 반영하겠습니다!

acceptor-gyu commented on be/src/main/java/yeonba/be/user/event/ProfilePhotoDeletionEvent.java:
```diff
@@ -0,0 +1,13 @@
+import lombok.RequiredArgsConstructor;
+import yeonba.be.user.entity.ProfilePhoto;
+
+@Getter
+@RequiredArgsConstructor
+public class ProfilePhotoDeletionEvent {
+
+    private final List<ProfilePhoto> profilePhotos;
+}
```
해당 클래스에 대해서 설명 부탁드립니다. 
+ EOF? EOL? 지켜주시기 바랍니다.

acceptor-gyu commented on be/src/main/java/yeonba/be/util/S3Service.java:
```diff
@@ -0,0 +1,94 @@
+
+        return uploadedProfilePhotosUrls;
+    }
+
+    public void deleteProfilePhotos(List<ProfilePhoto> profilePhotos) {
```
profilePhoto를 delete하는 로직이 필요없어 추가하지 않는 것으로 결론이 났던 것같은데 추가하신 이유가 무엇인가요?
+ 해당 로직 자체를 추가하면 안된다는 것은 아니지만 이유가 있더라도 한 번 결론이 났던 내용을 수정할 때는 논의 후에 수정하는 것이 좋을 것같은데 어떻게 생각하시나요?

Minjae-An replied:
불필요한 이하 작업 내역들 전부 삭제하겠습니다!
- 프로필 사진 삭제 로직
- 프로필 사진 삭제 이벤트 & 이벤트 리스너

acceptor-gyu commented on be/src/main/java/yeonba/be/util/JwtUtil.java:
```diff
@@ -0,0 +1,73 @@
+@Component
+public class JwtUtil {
+
+    private final Duration ACCESS_TOKEN_DURATION = Duration.of(8, ChronoUnit.HOURS);
+    private final Duration REFRESH_TOKEN_DURATION = Duration.of(10, ChronoUnit.DAYS);
```
제가 잘 몰라서 그러는데, ChronoUnit 클래스에 대해서 설명해주시면 감사하겠습니다.

Minjae-An replied:
`ChronoUnit`은 java 8에서 추가된 time 패키지에 속한 표준 시간 단위를 표현하는 데 사용되는 열거형입니다. 특정 라이브러리, 프레임워크가 제공하는 시간 단위를 사용해야만 하는 상황이 아니면 모든 상황을 커버할 수 있는 타입으로 알고 있습니다. 서로 다른 일시 단위인 access token과 refresh token의 유효 기간을 일관성 있게 표현하기 위해 사용하였습니다.

acceptor-gyu commented on be/src/main/java/yeonba/be/util/JwtUtil.java:
```diff
@@ -0,0 +1,73 @@
+    @Value("${JWT_SECRET}")
+    private String jwtSecret;
+
+    public String generateAccessToken(User user, Date issuedAt) {
+
```
`issued`라는 단어를 선택한 이유가 무엇인가요? 단어의 의미 자체는 적절하다고 생각하는데 기존부터 사용해와 저희에게 익숙한 `generated` 또는 `created`라는 단어를 사용하는 것은 어떨까요?

Minjae-An replied:
특정 라이브러리를 사용하는 코드에서는 라이브러리의 표현과 변수명을 맞추는 것이 일관성 있다고 생각했습니다. 
하지만 주신 의견이 더 일리 있다고 생각합니다. 수정하겠습니다!

acceptor-gyu replied:
아닙니다. 일리있습니다. 다만, jwt를 생성할 때 issued라는 단어를 라이브러리에서 사용하는데 그것에 맞춘다는 것 자체가 말 그대로 일관성보다는 의존성이라는 말이 어울릴 것같다는 생각이 듭니다.

acceptor-gyu commented on be/src/main/java/yeonba/be/util/JwtUtil.java:
```diff
@@ -0,0 +1,73 @@
+        return Date.from(instant);
+    }
+
+    private String generateUserJwt(
+        User user,
+        Date issuedAt,
+        Date expiredAt) {
```
해당 메서드에서만 파라미터를 한 줄에 하나씩 넣으신 이유가 무엇인가요?

Minjae-An replied:
너무 열이 길어서 조정하였는데 지금 보니 실수인 것 같습니다. 수정하겠습니다.

acceptor-gyu commented on be/src/main/java/yeonba/be/util/JwtUtil.java:
```diff
@@ -0,0 +1,73 @@
+                .setSigningKey(jwtSecret)
+                .parseClaimsJws(jwt);
+
+        } catch (SignatureException e) {
+            throw new IllegalStateException("유효하지 않은 JWT 시그니처입니다.", e);
+        } catch (ExpiredJwtException e) {
+            throw new IllegalStateException("만료된 JWT입니다.", e);
+        }
```
제가 이해를 잘못한 것인지 모르겠지만 서버에서 발생하는 에러는 전부 다 GeneralException으로 처리하자고 말씀하셨는데, 이 부분만 다르게  처리하신 이유는 무엇인가요?

추가로, 제 생각에는 저희 API를 제 3의 서비스가 사용하는 것이 아니기 때문에 구체적인 예외 메시지(시그니처가 잘못되었습니다, 만료된 JWT입니다 등)를 응답해줄 필요가 없을 것같은데 아래와 같이 처리하는 것에 대해서 어떻게 생각하시나요?
```suggestion
        } catch (SignatureException, ExpiredJwtException e) {
            throw new IllegalStateException("유효하지 않은 JWT입니다. 다시 로그인 해주세요.", e);
        }
```

Minjae-An replied:
유틸 기능은 어떤 기술에 의존함에 따라 변경 가능성이 잦은 기능이라고 생각했습니다. 그런 측면에서 유틸 기능에서 발생할 수 있는 예외를 도메인 연관 예외와 같은 형식으로 처리하는 것이 적절한 지 의문이 들어 일단 저렇게 정의하였습니다. 예외를 일괄적으로 처리하는 방식은 적절하다고 생각합니다. 반영하겠습니다.

acceptor-gyu replied:
util 패키지에서 발생했다면 UtilException과 같은 형태로 일관성 있게 맞춰도 차이가 없는 것같다는 생각이 드네요.
그리고 말씀하신 내용에서 `의존`이라는 단어를 사용하셨는데 의존함에 따라 변경이 잦아진다면 적절하게 역할을 분리한 것인지에 대해서 조금 더 고민을 해보는 것이 좋지 않을까 생각이 듭니다.

Minjae-An replied:
`UtilException`으로 일관성 있게 처리하는 것에 동의합니다. 반영하겠습니다.

acceptor-gyu commented on be/src/main/java/yeonba/be/util/JwtUtil.java:
```diff
@@ -0,0 +1,73 @@
+        Date issuedAt,
+        Date expiredAt) {
+
+        return Jwts.builder()
+            .setSubject(user.getEmail())
```
jwt는 token을 알고 있다면 누구나 내용을 알 수 있는데 subject에 사용자 이메일을 넣은 이유가 무엇인가요?
제 생각에는 특별한 장점이 없고 굳이 사용자의 이메일 정보를 노출시킬 필요가 없다고 생각해서 여쭤봅니다.

Minjae-An replied:
jwt의 `sub` 클레임은 토큰이 대상으로 하는 주체를 나타내는 용도로 사용한다고 알고 있습니다. 이에 관례적으로
이메일이 많이 쓰이는 것으로 알아 로직을 저런 형태로 구성하였습니다. 하지만, 주신 의견대로 
현재 jwt를 이용한 인증/인가 방식에서는 별다른 이점이 없는 것 같습니다. 수정하겠습니다.

acceptor-gyu commented on be/src/main/java/yeonba/be/user/service/JoinService.java:
```diff
@@ -0,0 +1,54 @@
+@Service
+@RequiredArgsConstructor
+public class JoinService {
+
+    private final ApplicationEventPublisher eventPublisher;
```
해당 클래스에 대한 설명 부탁드립니다.

acceptor-gyu commented on be/src/main/java/yeonba/be/user/service/UserService.java:
```diff
@@ -36,4 +73,100 @@ public UserProfileResponse getTargetUserProfile(long userId, long targetUserId)
+        user.updateProfilePhotos(profilePhotos);
+    }
+
+    @Transactional
+    public void saveUserPreference(User user, UserJoinRequest request) {
```
해당 메서드는 JoinService의 @Transactional이 붙어있는 join() 메서드에서 호출되는 메서드처럼 보이는데 @Transactional이 있는 이유는 무엇인가요?

Minjae-An replied:
- 스프링은 기존 트랜잭션 진행 중 다른 트랜잭션 로직이 실행되면 이 둘을 하나의 물리 트랜잭션(실제 DB 연결 트랜잭션)으로 다룹니다.
- 이런 최상위 물리 트랜잭션 하의 트랜잭션은 논리 트랜잭션으로 다뤄지며 기본적으로 하나의 논리 트랜잭션이라도 롤백될 경우 전체 물리 트랜잭션을 롤백합니다.
- 회원가입 과정에서 사용자 저장/프로필 사진 저장/선호 조건 저장은 한 트랜잭션 내에서 이뤄져야 하고 이들 중 한 작업이라도 롤백된다면 전체가 롤백되어야 하므로 스프링의 트랜잭션 관리를 이용하기 위해 하위 로직들에도 `@Transactional` 어노테이션을 사용했습니다.
- 한편, 해당 비즈니스 로직이 특정 상위 로직에 의해서만 이용되더라도 오해의 소지가 없도록 어떤 트랜잭션 하에서 실행되어야 하는 지 명시할 필요성이 있다고 생각합니다.

acceptor-gyu replied:
어떤 상황인지는 이해되지만, 무슨 말인지 잘 이해가 안됩니다.
제가 이해하는 바로는 2번째 문장과 3번째 문장이 서로 상반된 말씀을 하고 계시는 것같습니다.

일단 제 말씀을 드리자면,
상위 메서드에 `@Transactional`이 붙어있다면 해당 메서드에서 호출되는 메서드에는 `@Transactional`이 없더라도 전체가 하나의 트랜잭션으로 동작하는 것으로 알고 있습니다. 즉, JoinService의 join() 메서드에 `@Transactional`이 있기때문에 호출되는메서드들 중 하나라도 Rollback되면 전체가 Rollback 되는 것으로 알고 있습니다.

추가적으로, 이 부분을 이해하고 작성하신 거라면, `@Transactional` 남용으로 인해 발생하는 오버헤드에 대해서도 생각해보신 건가요?

Minjae-An replied:
제가 `@Transactional`의 사용에 대해 오인하고 있던 부분이 있었습니다. 수정토록 하겠습니다.

acceptor-gyu commented on be/src/main/java/yeonba/be/user/service/UserService.java:
```diff
@@ -36,4 +73,100 @@ public UserProfileResponse getTargetUserProfile(long userId, long targetUserId)
+        return userCommand.save(user);
+    }
+
+    @Transactional
+    public void saveProfilePhotos(User user, UserJoinRequest request) {
```
해당 메서드는 JoinService의 @transactional이 붙어있는 join() 메서드에서 호출되는 로직처럼 보이는데 @transactional이 있는 이유는 무엇인가요?

acceptor-gyu commented on be/src/main/java/yeonba/be/util/S3Service.java:
```diff
@@ -0,0 +1,94 @@
+     */
+    public List<String> uploadProfilePhotos(List<MultipartFile> profilePhotos, User user) {
+
+        if (!validateProfilePhotosExtensions(profilePhotos)) {
+            throw new IllegalArgumentException("jpg, jpeg, png 확장자 형식의 파일만 허용됩니다.");
+        }
```
기존과 다른 형식으로 예외처리를 한 이유가 무엇인가요?

Minjae-An replied:
앞선 jwt 관련 예외를 `IllegalArgumentException`으로 처리한 것과 같은 이유입니다.

Minjae-An committed:
refactor: 사용자 상세 프로필 응답 DTO 수정(#8)

- 사진 싱크로율 필드 타입 int로 수정
- 채형 필드 명세 오타 수정

Minjae-An committed:
fix: 불필요한 업로드된 사진 삭제 로직 제거(#8)

- 동일한 URL로 사진을 업로드하기에 이미 업로드된 사진 URL을 업데이트하는
  방식으로 데이터 무결성 보장 가능
- 이에 따라 불필요한 로직들 제거

Minjae-An committed:
refactor: jwt 생성 로직 수정(#8)

- issuedAt, 기존 서비스 명명 방식에 부합하고 더 친숙한 generatedAt으로
  수정
- 불필요한 사용자 이메일 sub 설정 제거

Minjae-An committed:
refactor: JWT 검증 예외, 일괄 처리하도록 로직 수정(#8)

Minjae-An committed:
refactor: 불필요한 트랜잭션 선언 제거(#8)

Minjae-An committed:
feat: 유틸 기능 연관 예외 정의(#8)

- 유효하지 않은 JWT
- 허용되지 않은 이미지 파일 확장자

Minjae-An committed:
refactor: 커스텀 예외 사용하도록 검증 로직 수정(#8)

Minjae-An committed:
refactor: 커스텀 예외 사용토록 로직 수정(#8)

Minjae-An committed:
Merge branch 'dev' into feat/#8-join-api

# Conflicts:
#	be/src/main/java/yeonba/be/exception/JoinException.java
#	be/src/main/java/yeonba/be/exception/LoginException.java
#	be/src/main/java/yeonba/be/login/controller/LoginController.java
#	be/src/main/java/yeonba/be/login/service/LoginService.java

Minjae-An committed:
feat: 인가 과정 제외 URL들 추가(#8)

인가 과정이 필요 없는 리소스 경로들 설정

Minjae-An committed:
refactor: 이미 사용하는 전화번호 검증 메서드명 수정(#8)

이미 사용하는 전화번호인지 확인하는 작업을 수행한다는 것을 잘 표현하기 위해 수정

Minjae-An committed:
feat: 이미 사용 중인 전화번호 검증 로직, 회원가입 로직에 추가(#8)

Minjae-An committed:
refactor: 병합과 메서드명 변경에 따른 수정(#8)

- 이미 사용 중인 전화번호 검증 메서드명 변경에 따라 코드 수정
- dev 브랜치 작업 내역 병합에 따른 수정

Minjae-An committed:
refactor: 병합에 따른 코드 수정(#8)

dev 브랜치 작업내역 병합에 따른 코드 수정

acceptor-gyu commented on be/src/main/java/yeonba/be/user/repository/UserQuery.java:
```diff
@@ -26,7 +26,7 @@ public User findByEmail(String email) {
 
-    public boolean existByPhoneNumber(String phoneNumber) {
+    public boolean isAlreadyUsedPhoneNumber(String phoneNumber) {
 
```
메서드 이름으로 짓기에 애매한 것같은데 `validateUsedPhoneNumber`는 어떠신가요?

Minjae-An replied:
수긍합니다. 일괄 수정했습니다!

acceptor-gyu commented on be/src/main/java/yeonba/be/exception/UtilException.java:
```diff
@@ -0,0 +1,34 @@
+
+        return reason;
+    }
+}
```
EOL 지켜주세요

acceptor-gyu commented on be/src/main/java/yeonba/be/user/service/UserService.java:
```diff
@@ -36,4 +73,106 @@ public UserProfileResponse getTargetUserProfile(long userId, long targetUserId)
+            preferredAnimal);
+        userPreferenceCommand.save(userPreference);
+    }
+}
```
EOL 지켜주세요

Minjae-An replied:
일괄 수정했습니다!

acceptor-gyu reviewed (approved):
issuedAt은 변수명을 변경하기로 하지 않았었나요??

validateJWT 메서드의 경우는 사용되는 곳이 없는데 이후 PR에 검증하는 곳을 추가하신 걸까요? 아니라면 제가 추가하고 PR 올릴게요.

Minjae-An committed:
refactor: jwt 검증 로직 삭제 및 로직 리팩토링(#8)

- 일단 사용하지 않는 jwt 검증 로직 삭제
- issuedAt 변수명, generatedAt으로 일괄 수정

Minjae-An committed:
refactor: EOL 추가, 코드 정렬(#8)

Minjae-An committed:
refactor: 사용 중인 데이터 검증 로직 이름 수정(#8)

사용 중인 데이터(이메일, 폰번호, 별명) 검증 로직 이름 일괄 수정

Minjae-An committed:
refactor: 메서드명 수정에 따른 사용 위치 수정(#8)

Minjae-An committed:
refactor: 메서드명 수정에 따른 사용 위치 수정(#8)

Minjae-An commented:
후속 PR에서 `validateJWT`를 사용하고 있습니다. 그 부분은 제가 추가토록 하겠습니다!
